core/offchain/impl/offchain_worker_impl.cpp

```diff
@@ -0,0 +1,340 @@
+#include "offchain/impl/offchain_local_storage.hpp"
+#include "runtime/executor.hpp"
+
+namespace kagome::offchain {
+
+  OffchainWorkerImpl::OffchainWorkerImpl(
+      const application::AppConfiguration &app_config,
+      std::shared_ptr<clock::SystemClock> clock,
+      std::shared_ptr<crypto::Hasher> hasher,
+      std::shared_ptr<storage::BufferStorage> storage,
+      std::shared_ptr<crypto::CSPRNG> random_generator,
+      std::shared_ptr<api::AuthorApi> author_api,
+      libp2p::Host &host,
+      std::shared_ptr<OffchainPersistentStorage> persistent_storage,
+      std::shared_ptr<runtime::Executor> executor,
+      const primitives::BlockHeader &header)
+      : app_config_(app_config),
+        clock_(std::move(clock)),
+        hasher_(std::move(hasher)),
+        random_generator_(std::move(random_generator)),
+        author_api_(std::move(author_api)),
+        host_(host),
+        persistent_storage_(std::move(persistent_storage)),
+        executor_(std::move(executor)),
+        header_(header),
+        log_(log::createLogger(
+            "OffchainWorker#" + std::to_string(header_.number), "offchain")) {
+    BOOST_ASSERT(clock_);
+    BOOST_ASSERT(hasher_);
+    BOOST_ASSERT(storage);
+    BOOST_ASSERT(random_generator_);
+    BOOST_ASSERT(author_api_);
+    BOOST_ASSERT(persistent_storage_);
+    BOOST_ASSERT(executor_);
+
+    auto hash = hasher_->blake2b_256(scale::encode(header_).value());
+    const_cast<primitives::BlockInfo &>(block_) =
+        primitives::BlockInfo(header_.number, hash);
+
+    local_storage_ =
+        std::make_shared<OffchainLocalStorageImpl>(std::move(storage));
+  }
+
+  outcome::result<void> OffchainWorkerImpl::run() {
+    BOOST_ASSERT(not current().has_value());
+
+    auto main_thread_func = [ocw = shared_from_this()] {
+      soralog::util::setThreadName("ocw." + std::to_string(ocw->block_.number));
+
+      current(*ocw);
+
+      SL_TRACE(ocw->log_,
+               "Offchain worker is started for block #{} hash={}",
+               ocw->block_.number,
+               ocw->block_.hash.toHex());
+
+      auto res = ocw->executor_->callAt<void>(
+          ocw->block_.hash, "OffchainWorkerApi_offchain_worker", ocw->header_);
+
+      current(std::nullopt);
+
+      if (res.has_failure()) {
+        SL_ERROR(ocw->log_,
+                 "Can't execute offchain worker for block #{} hash={}: {}",
+                 ocw->block_.number,
+                 ocw->block_.hash.toHex(),
+                 res.error().message());
+        return;
+      }
+
+      SL_DEBUG(ocw->log_,
+               "Offchain worker is successfully executed for block #{} hash={}",
+               ocw->block_.number,
+               ocw->block_.hash.toHex());
+    };
+
+    try {
+      std::thread(std::move(main_thread_func)).detach();
+    } catch (const std::system_error &exception) {
+      return outcome::failure(exception.code());
+    } catch (...) {
+      return outcome::failure(boost::system::error_code{});
```

Comment:
Why an empty error code?

Reply:
Because it is really unknown error. I'll try found mo relevant code, if any

Reply:
Appropriate system error code not found. Added common error for similar cases